### ISSUE TITLE
Add support for Temperature Scaling

### ISSUE THREAD

**Issue by Khanz9664** [Khanz9664/TrustLens#65]
# [Feature] Implement Temperature Scaling for Post-Hoc Calibration

###  Objective

Add support for **Temperature Scaling**, a widely used post-hoc calibration method, to improve probability reliability of classification models in TrustLens.

---

###  Background

TrustLens currently evaluates calibration quality (e.g., ECE), but does not provide built-in methods to **correct miscalibration**.

Temperature scaling is a simple yet effective technique that learns a single scalar parameter **T > 0**:

```

logits_scaled = logits / T

````

This scaling improves the calibration of predicted probabilities **without affecting model accuracy**.

---

###  Goal

Implement a `TemperatureScaler` class with a scikit-learn-like interface:

```python
from trustlens.calibrators import TemperatureScaler

scaler = TemperatureScaler()
scaler.fit(logits_val, y_val)   # Learns optimal temperature T
probs_cal = scaler.predict_proba(logits_test)
````

---

###  Suggested File Location

```
trustlens/calibrators/temperature.py
```

---

#### Core Implementation

* Learn scalar temperature **T > 0**
* Optimize **T** by minimizing **negative log-likelihood (NLL)**
* Apply scaling to logits before softmax
* Support **multi-class classification** (binary support is a bonus)

---

#### API Design

Implement the following interface:

```python
class TemperatureScaler:
    def fit(self, logits, y):
        """Learn optimal temperature T from validation logits."""
        ...

    def predict_proba(self, logits):
        """Return calibrated probabilities."""
        ...
```

---

#### Implementation Notes

* Use a stable optimization method (e.g., **LBFGS**)
* Ensure numerical stability when applying softmax
* Keep implementation lightweight and dependency-minimal (prefer NumPy / SciPy)

---

###  Integration (Important)

* Keep the implementation **modular** (no breaking changes)
* Ensure compatibility with existing calibration metrics (ECE, Brier score)
* Optional: allow usage before calling `analyze()`

---

###  Testing

Add tests to verify:

* Correct fitting of temperature parameter
* Output probabilities are valid (sum to 1)
* Calibration improves (optional but encouraged)
* Edge cases:

  * empty inputs
  * mismatched shapes
  * invalid logits

---

###  Optional Enhancements

* Compare calibration before vs after scaling (ECE improvement)
* Add example usage in `examples/`
* Extend to handle logits from different model types

---

###  Why This Matters

* Bridges **evaluation → correction** in the calibration pipeline
* Makes TrustLens more practical for real-world deployment
* Aligns with standard ML practices used in production systems

---

###  Getting Started

Comment below:

> "I’d like to work on this"

Happy to guide if needed 👍
